pss/pss.go

```diff
@@ -553,20 +555,58 @@ func (p *Pss) isSelfPossibleRecipient(msg *PssMsg, prox bool) bool {
 // SECTION: Message sending
 /////////////////////////////////////////////////////////////////////
 
-func (p *Pss) enqueue(msg *PssMsg) error {
+// Enqueue a message msg. Pending will be true for messages with a booked slot in the queue
+func (p *Pss) enqueue(msg *PssMsg, pending bool) error {
 	defer metrics.GetOrRegisterResettingTimer("pss.enqueue", nil).UpdateSince(time.Now())
 
 	outboxmsg := newOutboxMsg(msg)
-	select {
-	case p.outbox <- outboxmsg:
-		metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(len(p.outbox)))
-
-		return nil
-	default:
+	// We mutex between the calculation of the outbox capacity and the introduction in the channel so we make sure nobody
+	// add a message to the queue after checking outboxCapacity
+	mux := sync.Mutex{}
+	mux.Lock()
+	defer mux.Unlock()
+	pendingSize := p.getPending()
+	// Only allow defaultOutboxCapacity messages at most processed (both enqueued or being forwarded)
+	if pending || pendingSize < defaultOutboxCapacity { //If pending there is already a slot booked for this message
+		if !pending {
+			// We book a slot in the queue increasing pending messages and release it after successfully sending the message
+			p.increasePending()
+		}
+		select {
+		case p.outbox <- outboxmsg:
+			metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(len(p.outbox)))
+			return nil
+		default:
+		}
 	}
-
 	metrics.GetOrRegisterCounter("pss.enqueue.outbox.full", nil).Inc(1)
-	return errors.New("outbox full")
+	if pending {
+		return errors.New("unexpected outbox full for pending message!")
```

Comment:
Can this happen at all?

Reply:
Not really, if this happens is because we made something wrong. And in fact we did, because the mutext above is not protecting anything. Anyway we are going to study @zelig solution, so this will change.

Reply:
Do you want me to remove that check? I know it shouldn't be needed, it is just a safe check in case some modifications make that branch possible in the future (so we know what exactly happened)

Reply:
I was thinking maybe we should panic here.

Reply:
Ok added, `log.Crit` and `panic`, whenever this happens. Commit a6036b6.
This can be tested commenting mutex locking (lines 565-566) and running `TestMessageEnqueuedRaceCondition` test case.